torch/csrc/distributed/c10d/sequence_num.cpp

```diff
@@ -10,52 +10,61 @@ SequenceNum::SequenceNum() : num_(c10::nullopt) {}
 SequenceNum::SequenceNum(const uint64_t num) : num_(num) {}
 
 SequenceNum::SequenceNum(const SequenceNum& other) {
-  if (!other.isSet()) {
-    num_ = c10::nullopt;
-  } else {
-    num_ = other.get();
-  }
+  num_.withLock([&](auto& num) {
+    if (!other.isSet()) {
+      num = c10::nullopt;
+    } else {
+      num = other.get();
+    }
+  });
 }
 
 uint64_t SequenceNum::get() const {
-  std::lock_guard<std::mutex> lock(lock_);
-  return *num_;
+  uint64_t ret = 0;
+  num_.withLock([&](const auto num) {
+    TORCH_CHECK(num != c10::nullopt);
+    ret = *num;
+  });
+  return ret;
```

Comment:
Once https://github.com/pytorch/pytorch/pull/73794/files lands you should be able to directly return the return value of the `withLock<T>()` method.